supabase/functions/_script/backup_and_cleanup.ts

```diff
@@ -0,0 +1,78 @@
+import { createClient } from 'https://esm.sh/@supabase/supabase-js@^2.2.3';
+import 'https://deno.land/x/dotenv/load.ts';
+import { r2 } from '../_utils/r2';
+import type { Database } from '../_utils/supabase.types.ts';
+
+const supabaseUrl = Deno.env.get('SUPABASE_URL') || '***';
+const supabaseServiceRoleKey = Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') || '***';
+const backupFolder = 'backup'; 
+
+function useSupabase() {
+  const options = {
+    auth: {
+      autoRefreshToken: true,
+      persistSession: true,
+      detectSessionInUrl: false,
+    },
+  };
+  return createClient<Database>(supabaseUrl, supabaseServiceRoleKey, options);
+}
+
+async function backupAndDeleteOldEntries() {
+  const supabase = useSupabase();
+  const daysToKeep = 30; 
+  const pageSize = 1000; 
+
+  const currentDate = new Date();
+  const cutoffDate = new Date(currentDate.getTime() - daysToKeep * 24 * 60 * 60 * 1000);
+
+  let offset = 0;
+  let hasMoreData = true;
+
+  while (hasMoreData) {
+    const { data: oldEntries, error: queryError } = await supabase
+      .from<Database['public']['Tables']['app_versions']['Row']>('app_versions')
+      .select()
+      .lt('updated_at', cutoffDate.toISOString())
+      .range(offset, offset + pageSize - 1);
+
+    if (queryError) {
+      console.error('Error querying the Supabase table:', queryError.message);
+      return;
+    }
+
+    if (!oldEntries || oldEntries.length === 0) {
+      hasMoreData = false;
+      break;
+    }
+
+    const backupData = JSON.stringify(oldEntries);
+    const backupFilename = `backup-${currentDate.toISOString()}-page-${offset / pageSize}.json`;
+    const backupPath = `${backupFolder}/${backupFilename}`;
+
+    try {
+      await r2.upload(backupPath, new TextEncoder().encode(backupData));
+      console.log(`Backup saved to R2: ${backupPath}`);
+    } catch (backupError) {
+      console.error('Error saving backup to R2:', backupError);
+    }
+
+    const { error: deleteError } = await supabase
+      .from('app_versions')
+      .delete()
+      .lt('updated_at', cutoffDate.toISOString());
```

Comment:
If there is an error in uploading the backup data to R2. The delete query still executes, if it happens then data is lost forever.

It should not delete the data from the database. Instead, it should throw an error and the job stop executing and retry if necessary.